### PR TITLE
Add support for Simple Features

### PR DESCRIPTION
**Pull request by eivindhammers**
Supports `sf`, `sfc` and `sfg` class objects. `GEOMETRY` and
`GEOMETRYCOLLECTION` are not supported since they are not compatible
with shapefiles, and introduces ambiguities for QGIS modules.

If `load_output` is not `NULL`, and QGIS output is a vector layer,
`run_qgis` returns an object of the same class as the vector input.

`Spatial*` and `sf` objects can be supplied in the same parameter list.
In that case, a warning is given, and an `sf` object is returned (given
that `load_output` is not `NULL`)

### CODE REVIEW

**Comment by pat-s** [comment]
Thanks for your contribution @eivindhammers! @jannes-m and me are quite busy right now and we will check your PR as soon as we can :smiley:

**Comment by eivindhammers** [comment]
No worries. I have erroneously added three additional commits, so the PR is a bit messy. db630c8 is the one supposed to be included (but 790883a is fine too).

The two middle ones stems from an attempt to add GeoPackage support, in order to avoid the shapefile field name length limit, but it turned out more complicated than I expected, so I postponed it, hence the reverted commit.

**Comment by eivindhammers** [comment]
The PR merge has to wait until `sf` 0.3.5 is released on CRAN, since this is required to have `st_write` update layers.

**Comment by jannes-m** [comment]
Hi Eivind,

thanks again so much for your contribution! The support of `sf` is definitely essential! Ok, then we will wait for the next `sf` release before integrating you pull request!

All the best,

Jannes

**Comment by eivindhammers** [comment]
The "update" option of `st_write' unfortunately does not work as expected for Shapefiles; it does not replace existing layers, but appends. This is no good, so I've reverted c74ab62.

**Comment by pat-s** [comment]
@eivindhammers Thanks for your participation here! Although we are also short of unit tests in this package (and `run_qgis()` test still fail on travis), this `sf` implementation of yours is a good example for which an unit test would be really valuable. 

Especially when there are future changes within the `sf` package, we can quickly identify if the functionality might be broken by some updates. 

So it would be really great if you could add an unit test of within the `test-run-qgis.R` file under tests/testthat for your `run_qgis()` implementation. :grinning:

**Comment by eivindhammers** [comment]
Test added. There's probably a lot of r-package and Github etiquette/best practice that I'm not aware of, so thanks for the pointers.

File: DESCRIPTION [hunk 1]
```diff
@@ -35,5 +35,6 @@ SystemRequirements: Python (>= 2.7), QGIS (>= 2.12)
 Suggests: 
           knitr,
           rmarkdown,
-          testthat
+          testthat,
+          sf
```
Comment (pat-s) [inline DESCRIPTION]: Please insert sf dependency >= 0.3.5 and move to `Imports:`

File: R/processing.R [hunk 2]
```diff
@@ -831,13 +856,24 @@ run_qgis <- function(alg = NULL, params = NULL, check_params = TRUE,
                       overwrite_layer = TRUE)
       # return the result
       file.path(tmp_dir, paste0(names(params)[[i]], ".shp"))
+    } else if (any(tmp %in% c("sf", "sfc", "sfg"))) {
+      # st_write cannot currently replace layers, so file.remove() them
+      file.remove(list.files(path = tmp_dir, 
+                             pattern = names(params)[[i]],
+                             full.names = TRUE))
+      sf::st_write(params[[i]],
```
Comment (pat-s) [inline R/processing.R]: remove package tag `::` -> not used in R packages.  
Specify single functions in the documentation using `@importFrom sf st_write`

File: R/processing.R [hunk 3]
```diff
@@ -831,13 +856,24 @@ run_qgis <- function(alg = NULL, params = NULL, check_params = TRUE,
...
+                             full.names = TRUE))
+      sf::st_write(params[[i]], 
+                   dsn = file.path(tmp_dir, paste0(names(params)[[i]], ".shp")),
+                   driver = "ESRI Shapefile",
+                   quiet = TRUE)
+      # return the result
+      file.path(tmp_dir, paste0(names(params)[[i]], ".shp"))
     } else if (tmp == "RasterLayer") {
       fname <- file.path(tmp_dir, paste0(names(params)[[i]], ".asc"))
       raster::writeRaster(params[[i]], filename = fname, format = "ascii",
```
Comment (pat-s) [inline R/processing.R]: remove package tag `::` -> not used in R packages.  
Specify single functions in the documentation using `@importFrom raster writeRaster`

File: R/processing.R [hunk 4]
```diff
@@ -926,25 +962,30 @@ run_qgis <- function(alg = NULL, params = NULL, check_params = TRUE,
...
-      test <- try(expr = 
-                    rgdal::readOGR(dsn = dirname(fname),
-                                   layer = gsub("\\..*", "", 
-                                                basename(fname)),
-                                   verbose = FALSE),
-                  silent = TRUE
-      )
+      
+      if (input.sf == FALSE) {
+        test <- try(expr = rgdal::readOGR(dsn = dirname(fname),
```
Comment (pat-s) [inline R/processing.R]: remove package tag `::` -> not used in R packages.  
Specify single functions in the documentation using `@importFrom rgdal readOGR`

File: R/processing.R [hunk 5]
```diff
@@ -926,25 +962,30 @@ run_qgis <- function(alg = NULL, params = NULL, check_params = TRUE,
...
+                                          layer = gsub("\\..*", "", 
+                                                       basename(fname)),
+                                          verbose = FALSE),
+                    silent = TRUE)
+      } else {
+        test <- try(expr = sf::st_read(dsn = dirname(fname),
```
Comment (pat-s) [inline R/processing.R]: remove package tag `::` -> not used in R packages.  
Specify single functions in the documentation using `@importFrom sf st_read`

File: DESCRIPTION (hunk 1 above)
Comment (eivindhammers) [inline DESCRIPTION]: I though it made more sense for `sf` to be made voluntary, as a user will only ever encounter RQGIS' `sf` capabilities if he's already installed sf. But I'm happy to oblige.

File: R/processing.R (hunk 3 above)
Comment (eivindhammers) [inline R/processing.R]: It was like this before my PR (hence my `sf::` above). I'll fix both.

File: DESCRIPTION (hunk 1 above)
Comment (eivindhammers) [inline DESCRIPTION]: 0.3-5 not released and not actually required. Will import 0.3-4.

File: R/processing.R (hunk 3 above)
Comment (pat-s) [inline R/processing.R]: Jep, you're right. That was some leftover from our side.

File: DESCRIPTION (hunk 1 above)
Comment (pat-s) [inline DESCRIPTION]: That's right. However, you need to import the function in the doc to be able to use it and you can only import functions from packages which are listed under 'imports:' in the DESCRIPTION

**Pull request merged**